apps/site/lib/site/transit_near_me.ex

```diff
@@ -397,102 +355,308 @@ defmodule Site.TransitNearMe do
+      |> Enum.filter(fn %{predicted_schedule: predicted_schedule} ->
+        PredictedSchedule.has_prediction?(predicted_schedule)
+      end)
+      |> Enum.take(2)
+    else
+      predicted_schedules_with_time_data_and_crowding
+      |> Enum.take(2)
+      |> filter_predicted_schedules_with_time_data_and_crowding(nil)
+    end
+  end
 
-    {
-      predicted_schedule,
-      %{
-        delay: PredictedSchedule.delay(predicted_schedule),
-        scheduled_time: scheduled_time(predicted_schedule),
-        prediction: simple_prediction(predicted_schedule.prediction, route_type, now)
-      }
+  def filter_predicted_schedules_with_time_data_and_crowding(
+        [keep, %{predicted_schedule: %PredictedSchedule{prediction: nil}}],
+        _
+      ) do
+    # only show one schedule if the second schedule has no prediction
+    [keep]
+  end
+
+  def filter_predicted_schedules_with_time_data_and_crowding(
+        predicted_schedules_with_time_data_and_crowding,
+        _
+      ) do
+    predicted_schedules_with_time_data_and_crowding
+  end
+
+  @spec drop_predicted_schedules(predicted_schedule_with_time_data_and_crowding()) ::
+          time_data_with_crowding()
+  defp drop_predicted_schedules(%{
+         time_data: time_data,
+         crowding: crowding
+       }) do
+    %{
+      time_data: time_data,
+      crowding: crowding
     }
   end
 
+  @spec route_for_predicted_schedule_with_time_data_and_crowding(
+          predicted_schedule_with_time_data_and_crowding()
+        ) ::
+          Route.t()
+  defp route_for_predicted_schedule_with_time_data_and_crowding(%{
+         predicted_schedule: predicted_schedule
+       }),
+       do: PredictedSchedule.route(predicted_schedule)
+
+  @spec trip_for_predicted_schedule_with_time_data_and_crowding(
+          predicted_schedule_with_time_data_and_crowding()
+        ) ::
+          Trip.t() | nil
+  defp trip_for_predicted_schedule_with_time_data_and_crowding(%{
+         predicted_schedule: predicted_schedule
+       }),
+       do: PredictedSchedule.trip(predicted_schedule)
+
+  @spec headsign_for_predicted_schedule_with_time_data_and_crowding(
+          predicted_schedule_with_time_data_and_crowding()
+        ) ::
+          String.t() | nil
+  defp headsign_for_predicted_schedule_with_time_data_and_crowding(
+         predicted_schedule_with_time_data_and_crowding
+       ) do
+    case trip_for_predicted_schedule_with_time_data_and_crowding(
+           predicted_schedule_with_time_data_and_crowding
+         ) do
+      %Trip{headsign: headsign} ->
+        headsign
+
+      _ ->
+        nil
+    end
+  end
+
+  @spec crowding_for_predicted_schedule(PredictedSchedule.t()) :: Vehicle.crowding() | nil
+  defp crowding_for_predicted_schedule(predicted_schedule) do
+    with %Trip{id: trip_id} <- PredictedSchedule.trip(predicted_schedule),
+         route <- PredictedSchedule.route(predicted_schedule),
+         %Vehicle{crowding: crowding} <- Vehicles.Repo.trip(trip_id) do
```

Comment:
That's interesting. Being associated with the prediction is not sufficient to accept the vehicle, we also need to check that it's currently on the correct trip (to make sure it's not wrapping up its previous trip, for example), but that doesn't necessarily preclude this approach. 

Looking at it from the other direction, do we know for certain that if a vehicle is on a trip, there will always be a prediction for it for every stop? I could imagine a world in which there might not always be a prediction, but maybe it is guaranteed? I don't know enough to say, myself.

Reply:
Good questions - 

> we also need to check that it's currently on the correct trip (to make sure it's not wrapping up its previous trip, for example)

Hand-checking a few predictions/vehicles/trips, it looks like the prediction trip ID == the prediction's associated vehicle's trip ID, so if the prediction's on the correct trip (presumably it was in our existing implementation), then maybe the vehicle is too?

> Looking at it from the other direction, do we know for certain that if a vehicle is on a trip, there will always be a prediction for it for every stop? I could imagine a world in which there might not always be a prediction, but maybe it is guaranteed? 

Admittedly I don't know either. But how does the answer to this question impact this task at hand?

Reply:
I know for sure that there can be predictions for vehicles that are not on the same trip. I should have copied the note to this ticket, but you can see it on https://app.asana.com/0/555089885850811/1177064032714383. The example case here thinking of the 1 is say you are at Nubian station waiting at the first stop of the route. There can be a prediction for you, but the bus that is predicted to pick you up might be still finishing up its last trip, coming down the other way from Harvard. So the crowding level now doesn't affect you because all the current riders are going to get off before the bus lets you on.

As to whether there is a guaranteed prediction for every vehicle, that's important in thinking about how we get the vehicles and if we are missing any crowding data. I'm thinking about an example case where I'm waiting at a stop, a vehicle is scheduled to get to me in 10 minutes and there is a vehicle running that trip with crowding information, but for whatever reason, there's no prediction for when the bus will get to my stop—say the bus had to detour around some construction and the prediction engine got confused. If we pull vehicles based on being associated with a prediction, then we won't include that vehicle. But starting with all scheduled or predicted trips we would get this vehicle and be able to show the crowding information.

I feel like this conversation is digging into a lot of important details. Let me know if it would be better to talk through things in person.